### ISSUE TITLE
Eager Function Inputs cannot be Keras Symbolic Tensors error when using Intermediate layers in custom loss function

### ISSUE THREAD

**Issue by Ehsan1997** [tensorflow/tensorflow#40336]
I am using Colab with TF 2.2.

**Current behavior**
```
Inputs to eager execution function cannot be Keras symbolic tensors,
 but found [<tf.Tensor 'dense_3_1/Identity:0' shape=(100, 2) dtype=float32>
, <tf.Tensor 'dense_2_1/Identity:0' shape=(100, 2) dtype=float32>]
```
**Expected behavior**
Model trains successfully

**Standalone code to reproduce the issue**

1. Just copy the code from [here](https://github.com/keras-team/keras/blob/keras-2/examples/variational_autoencoder.py) to a colab cell,
2. Change keras to tensorflow.keras (using just keras works).
3. Code works with tf.config.experimental_run_functions_eagerly, but the problem is that it is very slow and sometimes causes OOM.

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/364f4b37d81a98b445bcae9061526193/40336.ipynb#scrollTo=Fe55PvC_VMGx) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/6762524676176dca9d3204a1e4f87ee1/40336-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by jvishnuvardhan** [comment]
@Ehsan1997 Can you please check this [tutorial](https://keras.io/examples/variational_autoencoder/) which is similar to your problem.  Thanks!

**Comment by Ehsan1997** [comment]
@jvishnuvardhan Thanks for your reply, but the tutorial you pointed to uses the same input as the output, which is not what I want.
I want some X (different from Y) to predict the Y.
Secondly what I don't understand is that why is this code running correctly for old keras but not for the latest tf2?

**Comment by jvishnuvardhan** [comment]
@Ehsan1997 In your code, you are using same `x_train` for X and Y. Check the `model.fit` function below.
`vae.fit(x_train, x_train, shuffle=True,epochs=epochs, batch_size=batch_size, validation_data=(x_test, x_test))
`
Regarding why `tf.keas` was not working when `keras` was working with the same code, in `tf.keras` model.fit runs in graph model by default. However, some ops in the custom loss function expects eager Tensors while Graph Tensors are provided. You can use `run_eagerly=True` to run without any issue (may take little more time). 

`vae.compile(optimizer='rmsprop', loss=VAE_Loss(),run_eagerly=True)`

Please check the following [guide](https://www.tensorflow.org/guide/keras/train_and_evaluate) on defining custom_loss class. Thanks!

**Comment by Ehsan1997** [comment]
@jvishnuvardhan Thanks for the reply.

Sorry for the confusion, but this code is the simplified version of what I am doing, I want a different image at output as to the one I am inputting.

Run eagerly option really affects the performance, I have already mentioned about this in the opening post of this issue. I tried using the sub-class method for defining the loss, the problem still persists. 

Is there any way to make only a portion of the code run eagerly, so that other code runs fast?

**Comment by tilakrayal** [comment]
Is this still an issue?
Could you please update TensorFlow to the latest stable version v.2.6 and let us know if you are facing the same error. Thanks!
